src/services/formatting/formatting.ts

```diff
@@ -438,6 +438,25 @@ namespace ts.formatting {
             }
         }
 
+        if (previousRange! && formattingScanner.getStartPos() >= originalRange.end) {

```

Comment:
why do we need the bang in `previousRange!` here?

Reply:
TypeScript is complaining that I’m using it before initialization, but it gets initialized in child functions. It’s complaining because the type annotation doesn’t include `undefined`, so it wants to prevent me from using it where it could be `undefined`. I think it actually could be `undefined` here since I don’t think the child functions are guaranteed to run here. Probably I should just annotate it with `| undefined`, but then I need to add bangs in multiple other places.

Reply:
I see, thanks for clarifying